src/backend/src/signer.rs

```diff
@@ -151,7 +159,52 @@ pub async fn btc_principal_to_p2wpkh_address(
 ///
 /// # Errors
 /// Errors are enumerated by: `TopUpCyclesLedgerError`
-#[allow(clippy::unused_async)] // TODO: Remove once the code is implemented
 pub async fn top_up_cycles_ledger(request: TopUpCyclesLedgerRequest) -> TopUpCyclesLedgerResult {
-    todo!("Add code that tops up the cycles ledger per this request: {request:?}")
+    let cycles_ledger = CyclesLedgerService(*CYCLES_LEDGER);
+    let account = Account {
+        owner: ic_cdk::id(),
+        subaccount: None,
+    };
+    let (ledger_balance,): (Nat,) = cycles_ledger
+        .icrc_1_balance_of(&account)
+        .await
+        .map_err(|_| TopUpCyclesLedgerError::CouldNotGetBalanceFromCyclesLedger)?;
+
+    if ledger_balance < request.threshold() {
+        // Decide how many cycles to keep and how many to send to the cycles ledger.
+        let own_canister_cycle_balance = Nat::from(ic_cdk::api::canister_balance128());
+        let to_send = own_canister_cycle_balance.clone() / Nat::from(100u32)
+            * Nat::from(request.percentage());
+        let to_retain = own_canister_cycle_balance.clone() - to_send.clone();
+
+        // Top up the cycles ledger.
+        let arg = DepositArgs {
+            to: account,
+            memo: None,
+        };
+        let to_send_128: u128 =
+            to_send.clone().0.try_into().unwrap_or_else(|err| {
+                unreachable!("Failed to convert cycle amount to u128: {}", err)
+            });
+        let (result,): (DepositResult,) =
+            call_with_payment128(*CYCLES_LEDGER, "deposit", (arg,), to_send_128)
+                .await
+                .map_err(|_| TopUpCyclesLedgerError::CouldNotTopUpCyclesLedger {
+                    available: own_canister_cycle_balance,
+                    tried_to_send: to_send.clone(),
+                })?;
+        let ledger_balance = result.balance;
+
+        Ok(TopUpCyclesLedgerResponse {
+            ledger_balance,
+            backend_cycles: to_retain,
```

Comment:
Do not know how precise this value should be but, `to_retain` is not exactly the remaining cycles because the call to the ledger consumed few cycles.

Reply:
I'm not looking for great precision.  Maybe when writing tests I'll tweak the numbers to make it precise but when monitoring cycle balances I'm not concerned with tiny fractions of a Frank, unlike someone I know. :-D

Reply:
Oki doki